File: plugins/cbre-il/skills/cbre-property-longlist/evals/percard_completeness_test.py
#!/usr/bin/env python3
"""percard_completeness_test.py - PER-PROPERTY completeness over the CURATED row set.

THE RULE, in one line: for each property individually, show every CURATED variable that has a
value for THAT property, and render no row for one it lacks. If option A has 10 filled curated
variables and option B has 15, A shows 10 rows and B shows 15. Never a page of tbd.

The row set itself is FIXED and hand-authored - the modal's three sections, Compare's row list
and the Flyover's spec block. A variable no curated row owns renders nowhere, however real its
value; it still reaches the broker through the Source Ledger and the Longlist workbook. That is
the second half of what this pins, because it is the half that silently rots: a chrome that
derives rows from a property's own key set will show whatever an interpretation record happened
to invent, under a machine-made label, in an order nobody chose.

WHAT THIS PINS. The modal and compare assertions EXECUTE the real chrome in node:vm (the .mjs
sibling), so they test behaviour, not prose. The Flyover assertion is structural - `slideHtml`
lives inside the view's IIFE and is not reachable from the sandbox - so it extracts the spec
block FROM the built template and pins both halves: each curated row is guarded so it omits when
absent, and no row is derived from the property's own keys.
Offline. Drives a real build.
"""
from __future__ import annotations
import json
import re
import shutil
import subprocess
import sys
import tempfile
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
HELPERS = ROOT / "helpers"
sys.path.insert(0, str(HELPERS))
import build_dashboard  # noqa: E402

PX = ("data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAQAAAC1HAwCAAAAC0"
      "lEQVR42mNkYPhfDwAChwGA60e6kgAAAABJRU5ErkJggg==")


def _canon(props):
    meta = {"client": "PerCard", "units": {"area": "sq m"},
            "hero": {"topbar_meta": "", "eyebrow": "", "title_html": "", "lede": "",
                     "footer_copyright": ""}}
    base = {"country": "CZ", "developer": "D", "city": "Bor", "status": "Available",
            "photo": PX, "gallery": [PX], "lat": 49.75, "lng": 12.77, "areaUnit": "sq m",
            "rentUnit": "€/sq m/yr"}
    return {"meta": meta, "pois": [], "regions": {},
            "properties": [dict(base, **p) for p in props]}


def main() -> int:
    fails = []

    def ck(ok, label):
        print(("  ok   " if ok else "  FAIL ") + label)
        if not ok:
            fails.append(label)

    with tempfile.TemporaryDirectory() as td:
        d = Path(td)
        cp, hp = d / "c.json", d / "b.html"
        cp.write_text(json.dumps(_canon([
            # RICH: canonical extras + two keys that exist in NO schema (the v22/v33 path)
            dict(id=1, park="Rich Park", warehouseArea=50000, warehouseRent="€60 / sq m / year",
                 warehouseRentVal=60, sprinklers="Yes", permitting="Consented",
                 landPrice="Not charged", incentives="6 months fit-out",
                 yardRent="€10,000 per annum", railSiding="Yes"),
            # LEAN: the same base, none of the extras
            dict(id=2, park="Lean Park", warehouseArea=30000, warehouseRent="€55 / sq m / year",
                 warehouseRentVal=55),
        ])), encoding="utf-8")
        build_dashboard.build(cp, hp)
        built = hp.read_text(encoding="utf-8")

        # ---------- Flyover: structural pin (slideHtml is closed over inside the view IIFE) ----------
        m = re.search(r'function slideHtml\(p, i\)\s*\{(.*?)\n  \}', built, re.S)
        ck(bool(m), "the built chrome still defines the Flyover slideHtml()")
        body = m.group(1) if m else ""
        ck("Object.keys(p)" not in body,
           "flyover: no row is derived from the property's own key set")
        ck(not any(t in body for t in ("DENY_FIELDS", "autoLabel", "LOCATOR_RE", "foCurated")),
           "flyover: none of the auto-attribute machinery is reachable from the slide")
        ck(body.count("isTbd(p.") >= 6,
           "flyover: every curated spec row is guarded, so it omits when the property lacks it")
        ck("specRow(T(" in body,
           "flyover: rows carry an i18n label, never a machine-derived one")
        ck("certStr(p)" in body,
           "flyover: the curated Certification row is present")

        # ---------- Modal + Compare: real execution ----------
        node = shutil.which("node") or r"C:\Users\TBaaij\nodejs\node.exe"
        mjs = Path(__file__).with_suffix(".mjs")
        if not Path(node).exists() and not shutil.which("node"):
            ck(False, "node is required to execute the chrome (install node or add it to PATH)")
        else:
            p = subprocess.run([node, str(mjs), str(hp)], capture_output=True, text=True,
                               encoding="utf-8", errors="replace")
            print((p.stdout or "").rstrip())
            if p.returncode != 0:
                print((p.stderr or "").rstrip())
            ck(p.returncode == 0, "the executed modal + compare assertions pass (see above)")

    print("STATUS:", "ALL-PASS" if not fails else "BLOCKED")
    return 1 if fails else 0


if __name__ == "__main__":
    raise SystemExit(main())

File: plugins/cbre-il/skills/cbre-property-longlist/evals/percard_completeness_test.mjs
// evals/percard_completeness_test.mjs - execute the REAL detailHTML() and compareHTML() from a
// BUILT dashboard and assert the per-property rule over the CURATED row set.
// Offline; no npm deps.  Usage: node percard_completeness_test.mjs <built_html_path>
import fs from 'node:fs';
import vm from 'node:vm';

const htmlPath = process.argv[2];
if (!htmlPath) { console.error('usage: node percard_completeness_test.mjs <built_html>'); process.exit(2); }
const html = fs.readFileSync(htmlPath, 'utf8');

const scripts = [...html.matchAll(/<script(?![^>]*\bsrc=)[^>]*>([\s\S]*?)<\/script>/gi)].map(m => m[1]);
const code = scripts.join('\n;\n') +
  '\n;\n__capture__("PROPS", typeof PROPS !== "undefined" ? PROPS : undefined);' +
  '\n__capture__("detailHTML", typeof detailHTML !== "undefined" ? detailHTML : undefined);' +
  '\n__capture__("compareHTML", typeof compareHTML !== "undefined" ? compareHTML : undefined);' +
  '\n__capture__("FIELD_PRESENT", typeof FIELD_PRESENT !== "undefined" ? FIELD_PRESENT : undefined);\n';

// See modal_render_test.mjs for why this sandbox is shaped the way it is (top-level const does
// not land on the context global; a has-always-true trap would otherwise shadow the intrinsics).
const sink = new Proxy(function () {}, {
  get: (_t, p) => (p === Symbol.toPrimitive || p === 'toString' || p === 'valueOf') ? () => '' : sink,
  apply: () => sink, construct: () => sink, has: () => true,
});
const target = { console };
for (const name of Object.getOwnPropertyNames(globalThis)) {
  if (!(name in target)) { try { target[name] = globalThis[name]; } catch { /* ignore */ } }
}
target.globalThis = target;
target.__capture__ = (name, val) => { target[name] = val; };
const ctx = vm.createContext(new Proxy(target, {
  get: (t, p) => (p in t ? t[p] : sink),
  has: () => true,
}));

try { vm.runInContext(code, ctx, { filename: 'built.inline.js' }); }
catch (e) { console.error('FAIL: template script threw during eval:', e && e.message); process.exit(1); }

const { PROPS: props, detailHTML, compareHTML } = target;
if (typeof detailHTML !== 'function' || typeof compareHTML !== 'function' || !Array.isArray(props)) {
  console.error('FAIL: could not capture detailHTML/compareHTML/PROPS from the built chrome');
  process.exit(1);
}

const fails = [];
const ck = (ok, label) => { console.log((ok ? '  ok   ' : '  FAIL ') + label); if (!ok) fails.push(label); };

const rich = props.find(p => p.park === 'Rich Park');
const lean = props.find(p => p.park === 'Lean Park');
if (!rich || !lean) { console.error('FAIL: fixture properties missing from PROPS'); process.exit(1); }

const rHtml = detailHTML(rich);
const lHtml = detailHTML(lean);

// --- the rule: a curated variable renders for the property that HAS it, and for no other ---
const CURATED = [
  ['Sprinklers', 'sprinklers'],
  ['Permitting', 'permitting'],
  ['Land price', 'landPrice'],
  ['Incentives', 'incentives'],
];
for (const [label, key] of CURATED) {
  ck(rHtml.includes(label), `modal: the rich property shows "${label}" (it carries ${key})`);
  ck(!lHtml.includes(label), `modal: the lean property does NOT show "${label}" - it has no value`);
}

const specCount = h => (h.match(/class="spec-k"/g) || []).length;
ck(specCount(rHtml) > specCount(lHtml),
   `modal: per-property row counts differ (rich ${specCount(rHtml)} > lean ${specCount(lHtml)})`);

// no empty rows anywhere: every rendered spec value must be a real value
const emptyRow = /<div class="spec-v">\s*(tbd|tbc|—|-|)\s*<\/div>/i;
ck(!emptyRow.test(rHtml), 'modal: the rich property renders no tbd/empty spec row');
ck(!emptyRow.test(lHtml), 'modal: the lean property renders no tbd/empty spec row');

// --- a variable no curated row owns renders NOWHERE, however real its value ---
const UNCURATED = ['Yard Rent', 'yardRent', 'Rail Siding', 'railSiding'];
for (const label of UNCURATED) {
  ck(!rHtml.includes(label), `modal: "${label}" has no curated row, so it renders nowhere`);
}

// --- Compare: a matrix, so rows are the union of what the CURATED list covers ---
const cHtml = compareHTML([rich, lean]);
ck(cHtml.includes('Land price'), 'compare: carries the curated Land price row');
ck(cHtml.includes('Sprinklers'), 'compare: carries the curated Sprinklers row');
ck(cHtml.includes('Incentives'), 'compare: carries the curated Incentives row');
for (const label of UNCURATED) {
  ck(!cHtml.includes(label), `compare: "${label}" has no curated row, so it renders nowhere`);
}

// the v9 gate still holds: a curated row NO property carries appears nowhere
ck(!cHtml.includes('Never Stated'), 'compare: a field no property carries is still dropped');
ck(!rHtml.includes('Never Stated'), 'modal: likewise');

console.log('STATUS:', fails.length ? 'BLOCKED' : 'ALL-PASS');
process.exit(fails.length ? 1 : 0);
